File: hamster-back/library-core/src/main/java/com/librarycore/loan/Loan.java
package com.librarycore.loan;

import java.time.LocalDate;

public record Loan(
        String loanId,
        String userId,
        String inventoryId,
        LocalDate loanDate,
        LocalDate dueDate,
        LoanStatus status
) {

    public Loan returned() {
        return new Loan(
                loanId,
                userId,
                inventoryId,
                loanDate,
                dueDate,
                LoanStatus.RETURNED
        );
    }

    public boolean isOverdue(LocalDate today) {
        return today.isAfter(dueDate) && status == LoanStatus.LOANED;
    }
}

File: hamster-back/library-core/src/main/kotlin/com/librarycore/loan/Loan.kt
package com.librarycore.loan

import java.time.LocalDate


@JvmRecord
data class Loan(
    val loanId: String,
    val userId: String,
    val inventoryId: String,
    val loanDate: LocalDate,
    val dueDate: LocalDate,
    val status: LoanStatus,
) {
    fun returned(): Loan {
        return Loan(
            loanId,
            userId,
            inventoryId,
            loanDate,
            dueDate,
            LoanStatus.RETURNED
        )
    }

    fun isOverdue(today: LocalDate): Boolean {
        return today.isAfter(dueDate) && status == LoanStatus.LOANED
    }
}
